### PR TITLE
[WIP] add `pthread` as alpaka dependency

### PR DESCRIPTION
**Pull request by psychocoderHPC**
`std::future` needs `-pthread` at compile time to work correctly

see cupla issue: https://github.com/ComputationalRadiationPhysics/cupla/pull/128#issuecomment-545078917

We already passed `pthread` to the linker but we need also activate `pthread` usage for the compiler.
The implementation expresses the dependancy pthread by using `FIND_PACKAGE(Threads REQUIRED)`

### CODE REVIEW

File: alpakaConfig.cmake
Comment (sbastrakov) [inline alpakaConfig.cmake]: `pthread` should be `pthreads` I guess.

File: alpakaConfig.cmake
Comment (sbastrakov) [inline alpakaConfig.cmake]: Started discussing offline: I am not sure this should be `INTERFACE`, think `PRIVATE` instead since we only use `std::future` internally and this is not part of alpaka API.

Comment (psychocoderHPC) [reply alpakaConfig.cmake]: CMake is saying:
```
  INTERFACE library can only be used with the INTERFACE keyword of
  target_link_libraries
Call Stack (most recent call first):

```
Since the target alpaka is an `INTERFACE`  so we need to use `INTERFACE` here.

Comment (psychocoderHPC) [reply alpakaConfig.cmake]: but there is one space to much after `INTERFACE`, this I will fix after but like to wait for other reviews and suggestions first.

File: alpakaConfig.cmake
Comment (BenjaminW3) [inline alpakaConfig.cmake:241]: Are we sure that this is required on Mac as well?

Comment (psychocoderHPC) [reply alpakaConfig.cmake:241]: No I am not sure. I expect it since MAC is also using Pthreads for the system threads.
IMO it is not a big deal if we require on MAC pthreads too.

Comment (psychocoderHPC) [reply alpakaConfig.cmake:241]: before we added pthreads for MAC to the linker: https://github.com/ComputationalRadiationPhysics/alpaka/pull/866/files#diff-cd45f124e237a4bd8df18ffab83a5664L224
So this PR will only additionally add it to the compiler too.

File: alpakaConfig.cmake
Comment (sbastrakov) [reply alpakaConfig.cmake]: I see, thanks for explanations!

**Comment by BenjaminW3** [comment]
HIP does not seem to like the `-pthread` option.

**Comment by BenjaminW3** [comment]
I am still wondering, why this is not a problem for the alpaka CI and other projects using alpaka?

**Comment by psychocoderHPC** [comment]
> HIP does not seem to like the `-pthread` option.

Yes `nvcc` also do not know this flag because it must be passed to the host compiler

**Comment by psychocoderHPC** [comment]
> I am still wondering, why this is not a problem for the alpaka CI and other projects using alpaka?

I am not sure why but clearly need it on different systems. I think it is a side effect because alpaka is enabling all backends and this flag is automatically added with `OpenMP`

File: alpakaConfig.cmake
Comment (ax3l) [inline alpakaConfig.cmake:246]: maybe even:
```cmake
if(NOT DEFINED THREADS_PREFER_PTHREAD_FLAG)
    SET(THREADS_PREFER_PTHREAD_FLAG TRUE)
endif()
```
so users can overwrite it from the command-line.

**Comment by BenjaminW3** [comment]
Maybe source of related information: https://github.com/google/googletest/pull/2483

**Comment by ax3l** [comment]
That diff looks more like a bunch of pre-CMake 3.1 work-arounds to me, otherwise identical :)

**Comment by ienorand** [comment]
As far as I've understood whilst digging into this for googletest, nvcc needs `-Xcompiler -pthread` instead, see https://gitlab.kitware.com/cmake/cmake/issues/18008 .

(quoting myself) https://github.com/google/googletest/pull/2483#issuecomment-538761499
> So reading up on CUDA a bit it seems like -pthreads should be passed using -Xcompiler for nvcc as a special case, which seems to have been implemented only relatively recently in CMake according to https://gitlab.kitware.com/cmake/cmake/merge_requests/2512. Which seems to have ended up in Cmake 3.13.0.

Since you are using Cmake 3.11.4, that might be why you are seeing the nvcc issues still?

**Comment by ax3l** [comment]
Thank you for the feedback!
Alright, so if we bump our build dependencies to [CMake 3.13.0+](https://gitlab.kitware.com/cmake/cmake/commit/d81c4db489d773e4a1ce6be49eb9c41582e68ae3) ([CMake 3.14.0+ branch](https://gitlab.kitware.com/cmake/cmake/commit/69275d3b6a9787268f7bd80bb24d8e4cbb18710e)) or the potentially the backported [CMake 3.12.4](https://gitlab.kitware.com/cmake/cmake/commit/d17755f9cf583387fab83ad5744b93c746aa5994) we should be able to solve this as well.

**Comment by psychocoderHPC** [comment]
> Thank you for the feedback!
> Alright, so if we bump our build dependencies to [CMake 3.13.0+](https://gitlab.kitware.com/cmake/cmake/commit/d81c4db489d773e4a1ce6be49eb9c41582e68ae3) ([CMake 3.14.0+ branch](https://gitlab.kitware.com/cmake/cmake/commit/69275d3b6a9787268f7bd80bb24d8e4cbb18710e)) or the potentially the backported [CMake 3.12.4](https://gitlab.kitware.com/cmake/cmake/commit/d17755f9cf583387fab83ad5744b93c746aa5994) we should be able to solve this as well.

I will check if CMake 3.12.4  solved the issue, since we using still the old why to use nvcc (not the CMake native compiler) I am not sure if your linked backport solves it.

**Comment by ax3l** [comment]
Yes, I was wondering that as well - we are not using the CUDA language feature for targets yet (since we also support `clang -x cuda` as compiler).

**Comment by psychocoderHPC** [comment]
From our CI it looks like only hip is affected by this problem. I try to solve this issue but have the problem that I have currently no access to a working nvidia system with hip support :-(

**Comment by ax3l** [comment]
Doesn't Summit have official HIP modules to run on Nvidia?

**Comment by j-stephan** [comment]
I'm not entirely sure why but this will also be fixed once #921 is merged. The CI from #921 reports the following line for HIP + nvcc + `Threads::Threads` (correctly choosing `-lpthread` instead of `-pthread`:

```
Feb 24 13:18:44 [  2%] Linking HIP executable bufferCopy

Feb 24 13:18:44 cd /home/travis/build/ComputationalRadiationPhysics/alpaka/build/example/bufferCopy && /home/travis/CMake/bin/cmake -E cmake_link_script CMakeFiles/bufferCopy.dir/link.txt --verbose=1

Feb 24 13:18:44 /home/travis/hip/bin/hipcc_cmake_linker_helper /opt/rocm/hcc    CMakeFiles/bufferCopy.dir/src/bufferCopy_generated_bufferCopy.cpp.o -o bufferCopy  -lpthread /usr/lib/x86_64-linux-gnu/librt.so /home/travis/hip/hiprand/lib/libhiprand-d.so -lcudart
```

**Comment by psychocoderHPC** [comment]
> I'm not entirely sure why but this will also be fixed once #921 is merged. The CI from #921 reports the following line for HIP + nvcc + `Threads::Threads` (correctly choosing `-lpthread` instead of `-pthread`:
> 
> ```
> Feb 24 13:18:44 [  2%] Linking HIP executable bufferCopy
> 
> Feb 24 13:18:44 cd /home/travis/build/ComputationalRadiationPhysics/alpaka/build/example/bufferCopy && /home/travis/CMake/bin/cmake -E cmake_link_script CMakeFiles/bufferCopy.dir/link.txt --verbose=1
> 
> Feb 24 13:18:44 /home/travis/hip/bin/hipcc_cmake_linker_helper /opt/rocm/hcc    CMakeFiles/bufferCopy.dir/src/bufferCopy_generated_bufferCopy.cpp.o -o bufferCopy  -lpthread /usr/lib/x86_64-linux-gnu/librt.so /home/travis/hip/hiprand/lib/libhiprand-d.so -lcudart
> ```

offline discussed: `-lpthread` and `-pthread` are different things and `-pthread` should be a compiler flag. We will look into this issue after the cmake refactoring is merged.

**Comment by BenjaminW3** [comment]
@psychocoderHPC Is this now resolved?

**Comment by j-stephan** [comment]
Nope, this will need further work.

**Comment by psychocoderHPC** [comment]
Yes this must still be solved. I will work with @j-stephan on that part but it will take some time.

File: alpakaConfig.cmake
Comment (ax3l) [inline alpakaConfig.cmake:1103]: just generally: the target name `alpaka` does not need to be placed in quotes `""`.
Maybe remove everywhere in a follow-up.

**Comment by BenjaminW3** [comment]
2 of the 3 new lines from this PR are now already part of [`alpakaCommon.cmake`](https://github.com/alpaka-group/alpaka/blob/develop/cmake/alpakaCommon.cmake#L141-L142)
* `FIND_PACKAGE(Threads REQUIRED)`
* `TARGET_LINK_LIBRARIES("alpaka" INTERFACE Threads::Threads)`

Only `SET(THREADS_PREFER_PTHREAD_FLAG TRUE)` is not there yet. Is this required?
Can this PR be closed as it is already done?

**Comment by psychocoderHPC** [comment]
> Only `SET(THREADS_PREFER_PTHREAD_FLAG TRUE)` is not there yet. Is this required?
> Can this PR be closed as it is already done?

The is the most important line of this PR because this will set `-pthread` for compiling which is activating some defines which are required.
Please keep this PR open. 

Note: I am not available next week.

**Comment by j-stephan** [comment]
What is the state here?

**Comment by psychocoderHPC** [comment]
This PR is strongly required but I never found time to work on it.

File: alpakaConfig.cmake
Comment (bernhardmgruber) [inline alpakaConfig.cmake:241]: IIRC you can use `find_package` and `target_link_library` with `Threads` just fine with MSVC on Windows. No special handling is required. Please remove the `IF(NOT MSVC)` here and down below.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

**Comment by j-stephan** [comment]
Closed by #1420.

**Pull request closed without merging**
